### PR TITLE
Rejection sampling variational inference

### PR DESCRIPTION
**Pull request by cavaunpeu**
WIP. Addresses [#379](https://github.com/blei-lab/edward/issues/379).

### CODE REVIEW

**Comment by dustinvtran** [comment]
Thanks for working on this! Ping me whenever you'd like some feedback.

**Comment by cavaunpeu** [comment]
Will do! It's very WIP for now. I want to get a thing working, then clean it up considerably.

**Comment by cavaunpeu** [comment]
@dustinvtran, @naesseth

Hi. A few updates/questions:

- I implemented a `KucukelbirOptimizer` as given in Equation 9 of the [paper](http://arxiv.org/abs/1610.05683). (Too lazy to search for its real name, I called it this.) I put it in `edward/optimizers/sgd.py` and conformed its API to that of the other `optimizer` types that `VariationalInference.initialize` expects; it's a simple interface with an `apply_gradients` method. Eventually, this is something to robust-ify and merge upstream into Tensorflow, I'd think (if it's not there already). I've added an integration test as well.

- My next step is to implement a `ReparameterizedRejectionSampler` object. This doesn't need to know anything about VI. The motivation is the ability to test it separately.

After this, everything should be on the right track.

Thus far, I've worked on `build_rejection_sampling_loss_and_gradients`, as a corollary to `build_reparam_loss_and_gradients`. Would the gradients given in the paper easily extend to a
- `build_rejection_sampling_kl_loss_and_gradients`
- `build_rejection_sampling_entropy_loss_and_gradients`

as well? I haven't taken the time to think this through.

How does this all sound? Thanks!

File: edward/optimizers/sgd.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  """
+
+  def __init__(self, t, delta, eta, s_n, n):
+    self.t = t
+    self.delta = delta
+    self.eta = eta
+    self.s_n = s_n
+    self.n = n
+
+  def apply_gradients(self, grads_and_vars, global_step=None):
```
Comment (cavaunpeu) [inline edward/optimizers/sgd.py]: @dustinvtran 

I'd quite appreciate if you could glance at this method as well, as my integration test passes on some days and fails on others — with 0 changes to my code. Promise 🤞.

**Comment by cavaunpeu** [comment]
One more, and more for @dustinvtran:

It looks to me like you don't explicitly sample-via-reparameterization from `q(z)` in, for instance,
 [build_reparam_loss_and_gradients](https://github.com/blei-lab/edward/blob/6419751d1d849c84c502e5ff3f7249b9bbc7b3aa/edward/inferences/klqp.py#L650).

Why is this? And I missing something? Does it happen under the hood?

**Comment by dustinvtran** [comment]
Great work! Will take a look soon. To address questions I can do immediately:

> It looks to me like you don't explicitly sample-via-reparameterization from q(z) in, for instance,
build_reparam_loss_and_gradients. Why is this? And I missing something? Does it happen under the hood?

Take a look at how sampling and `ReparameterizationType` works in Distributions (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/distributions/distribution.py#L200). The idea is that certain sampling methods admit gradients with respect to distribution parameters, others don't, and this is demarcated with a type.

> I implemented a KucukelbirOptimizer as given in Equation 9 of the paper. (Too lazy to search for its real name, I called it this.) 

You can implement the optimizer by just calling RMSPropOptimizer and with a decaying learning rate (https://github.com/blei-lab/edward/blob/master/edward/inferences/variational_inference.py#L89). No need for a new optimizer object.

**Comment by cavaunpeu** [comment]
Hey @dustinvtran. I've:
- Removed the "Alp" sampler, and specified `'rmsprop'` with a decaying learning rate.
- Unit-tested a `GammaRejectionSampler` object.
- Unit-tested `build_rejection_sampling_loss_and_gradients`.

In the latter two, I pinned results to those computed in [this notebook](https://github.com/blei-lab/ars-reparameterization/blob/master/gamma/demo.ipynb) and vetted them thoroughly.

My gradients are still exploding. Might you have a chance to give some 👀 this week?

**Comment by williamabrwolf** [comment]
Hey @dustinvtran. Following up :)

**Comment by dustinvtran** [comment]
Apologies for the delay! Busy for ICML stuff due this Friday. Maybe ping me this weekend? :)

**Comment by williamabrwolf** [comment]
Ah, no worries! Will do.

The PR is in a good state, I think. Just need to close the gap between "unit-testing 1 iteration of RSVI gradient calculations passes" and "it just works". I'm guessing there's some slight "magic" with learning rates, supports, or the like that I'm missing. This is my hunch.

File: edward/inferences/inference.py [hunk 2]
```diff
@@ -144,6 +143,7 @@ def run(self, variables=None, use_coordinator=True, *args, **kwargs):
 
     for _ in range(self.n_iter):
       info_dict = self.update()
+      print(info_dict)
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: rm?

File: edward/inferences/inference.py [hunk 3]
```diff
@@ -123,7 +123,6 @@ def run(self, variables=None, use_coordinator=True, *args, **kwargs):
         Passed into `initialize`.
     """
     self.initialize(*args, **kwargs)
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: add back newline? unrelated to PR

File: edward/inferences/klpq.py [hunk 4]
```diff
@@ -32,7 +32,7 @@ class KLpq(VariationalInference):
 
   with respect to $\\theta$.
 
-  In conditional inference, we infer $z` in $p(z, \\beta
+  In conditional inference, we infer $z$ in $p(z, \\beta
```
Comment (dustinvtran) [inline edward/inferences/klpq.py]: This is unrelated to this PR. Can you make a new PR to fix this?

File: edward/inferences/klqp.py [hunk 5]
```diff
@@ -1127,3 +1184,90 @@ def build_score_rb_loss_and_gradients(inference, var_list):
...
+      tf.summary.scalar("loss/reparam_objective", rep,
+                        collections=[inference._summary_key])
+      tf.summary.scalar("loss/correction_term", cor,
+                        collections=[inference._summary_key])
+      tf.summary.scalar("loss/q_entropy", q_entropy,
+                        collections=[inference._summary_key])
+      tf.summary.scalar("loss/reg_penalty", reg_penalty,
+                        collections=[inference._summary_key])
+
+    g_rep = tf.gradients(rep, var_list)
```
Comment (dustinvtran) [inline edward/inferences/klqp.py]: Can you explain why you need the multiple gradient calls and not just one? This seems inefficient.

**Comment by cavaunpeu** [comment]
Hey @dustinvtran. Thanks for the feedback. I left this in "debug" state (print statements, etc.), as it doesn't yet work. Specifically, the main integration test, `_test_poisson_gamma`, does not pass: the gradients explode to `np.nan`.

Any insight as to why this might be? I'm guessing I'm lacking some expertise in getting this to work. NB: the "single pass gradient computation" integration test, `_test_build_rejection_sampling_loss_and_gradients`, does pass (and has been thoroughly cross-checked with the [Blei-lab notebook](https://github.com/blei-lab/ars-reparameterization/blob/master/gamma/demo.ipynb)).

Also, I think there are other code-organizational considerations to be made before readying this for merge. For instance, what happens when we have > 1 latent variables, where some require a rejection sampler and others don't?

In short: I think I need some help :)

**Comment by cavaunpeu** [comment]
Ping, @dustinvtran :) What's the best way forward? Have a moment to review in the coming days? Anyone else I should reach out to?

Cheers :)

**Comment by dustinvtran** [comment]
Not sure if I know enough about the algorithm to help unfortunately. What happens if you try 1000 samples per iteration? Maybe @naesseth can reply?

**Comment by slinderman** [comment]
Hi @williamabrwolf, I’m traveling now but I’ll be back in NYC mid March. Happy to meet then if you’d like. I’ll have limited cycles between now and then but I’ll try to take a look at the code here. Perhaps Christian can also help out in the meantime. Glad to see you’re integrating this into Edward!

**Comment by williamabrwolf** [comment]
@slinderman mid-March works great. Grateful for the help, and happy to meet then if we haven't resolved. williamabrwolf@gmail.com is me. Cheers :).

**Comment by naesseth** [comment]
I'll try and take a look early next week, currently traveling.

**Comment by williamabrwolf** [comment]
@slinderman @naesseth hey all. back in NYC? would love to meet for 2 hours and get this to a place where we can soon merge.

**Comment by williamabrwolf** [comment]
Hey @slinderman, @naesseth. Following up :). Back in the city? Happy to travel to you to make this happen.

Cheers!

**Comment by slinderman** [comment]
Hey Will,

I'm around this week.  Send me an email directly and we'll set up a time to
meet.

Best,
Scott

File: edward/inferences/klqp.py [hunk 6]
```diff
@@ -1127,3 +1184,90 @@ def build_score_rb_loss_and_gradients(inference, var_list):
...
+    g_cor = tf.gradients(cor, var_list)
+    g_entropy = tf.gradients(q_entropy, var_list)
+
+    grad_summands = zip(*[g_rep, g_cor, g_entropy])
```
[the remark this answers is not in the document]
Comment (cavaunpeu) [reply edward/inferences/klqp.py]: "With a reasonably small step size, maybe 100 epochs." --> i.e. it should pass "with a reasonably small step size, and run for maybe 100 epochs."
